imap_processing/codice/utils.py

```diff
@@ -57,3 +96,203 @@ class CoDICECompression(IntEnum):
+        and three_d_collapsed value and others.
+
+    Returns
+    -------
+    tuple[np.ndarray, np.ndarray]
+        (center_times, delta_times).
+    """
+    # If Lo sensor
+    if view_tab_obj.sensor == 0:
+        # Lo sensor, we need to set spins to be constant.
+        # 32 half spins makes full 16 spins for all non direct event products.
+        # But Lo direct event's spins is also 16 spins. Because of that, we can use
+        # the same calculation for all Lo products.
+        num_spins = 16.0
+    # If Hi sensor and Direct Event product
+    elif view_tab_obj.sensor == 1 and view_tab_obj.apid == CODICEAPID.COD_HI_PHA:
+        # Use constant 16 spins for Hi PHA
+        num_spins = 16.0
+    # If Non-Direct Event Hi product
+    else:
+        # Use 3d_collapsed value from LUT for other Hi products
+        num_spins = view_tab_obj.three_d_collapsed
+
+    # Units of 'spin ticks', where one 'spin tick' equals 320 microseconds
+    spin_period_ns = spin_period.astype(np.float64) * 320 * 1e3  # Convert to ns
+    delta_times = (num_spins * spin_period_ns) / 2
+    # subseconds need to converted to seconds using this formula per CoDICE team:
+    #   subseconds / 65536 gives seconds
+    center_times_seconds = (
+        acq_start_seconds + acq_start_subseconds / 65536 + (delta_times / 1e9)
+    )
+
+    return met_to_ttj2000ns(center_times_seconds), delta_times
+
+
+def calculate_acq_time_per_step(low_stepping_tab: dict) -> np.ndarray:
+    """
+    Calculate acquisition time per step from low stepping table.
+
+    Parameters
+    ----------
+    low_stepping_tab : dict
+        The low stepping table from the SCI-LUT JSON.
+
+    Returns
+    -------
+    np.ndarray
+        Array of acquisition times per step of shape (num_esa_steps,).
+    """
+    # These tunable values are used to calculate acquisition time per step
+    tunable_values = low_stepping_tab["tunable_values"]
+
+    # pre-calculate values
+    sector_time = tunable_values["spin_time_ms"] / tunable_values["num_sectors_ms"]
+    sector_margin_ms = tunable_values["sector_margin_ms"]
+    dwell_faction = tunable_values["dwell_fraction_percentage"]
+    min_hv_settle_ms = tunable_values["min_hv_settle_ms"]
+    max_hv_settle_ms = tunable_values["max_hv_settle_ms"]
+    num_steps_data = np.array(
+        low_stepping_tab["num_steps"].get("data"), dtype=np.float64
+    )
+    # Total non-acquisition time is in column (BD) of science LUT
+    total_non_acq_time = float(sector_time) * (100.0 - dwell_faction) / 100.0
+
+    # Calculate HV settle time per step not adjusted for Min/Max.
+    # It's in column (BF) of science LUT.
+    non_adjusted_hv_settle_per_step = (
+        total_non_acq_time - sector_margin_ms
+    ) / num_steps_data
+    hv_settle_per_step = np.minimum(
+        np.maximum(non_adjusted_hv_settle_per_step, min_hv_settle_ms), max_hv_settle_ms
+    )
+
+    # acquisition time per step in seconds
```

Comment:
Do you need to divide sector_margin_ms by 1000? Or update this comment to be in ms.